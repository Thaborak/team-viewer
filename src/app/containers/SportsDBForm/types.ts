import { Team } from 'types/Team';
/* --- STATE --- */
export interface SportsDbFormState {
  teamname: string;
  loading: boolean;
  error?: TeamErrorType | null;
  teams: Team[];
}

export enum TeamErrorType {
  RESPONSE_ERROR = 1,
  TEAM_NOT_FOUND = 2,
  TEAMNAME_EMPTY = 3,
  TEAM_HAS_NO_DATA = 4,
  SPORTSDB_RATE_LIMIT = 5,
}

/* 
  If you want to use 'ContainerState' keyword everywhere in your feature folder, 
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = SportsDbFormState;
