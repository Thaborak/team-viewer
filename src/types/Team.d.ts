// Copied from GraphQL wrapper for TheSportsDB

export interface Team {
  idTeam: string;
  strTeam: string;
  strTeamShort: string;
  strAlternate: string;
  intFormedYear: string;
  strSport: string;
  strKeywords: string;
  strDescriptionEN: string;
  strDescriptionES: string;
  strGender: string;
  strCountry: string;
  strTeamBadge: string;
  strTeamJersey: string;
  strTeamLogo: string;
  strTeamBanner: string;
  strWebsite: string;
  strManager: string;
  strStadium: string;
  strStadiumThumb: string;
  strStadiumDescription: string;
  intStadiumCapacity: number;
  strStadiumLocation: string;
  strTwitter: string;
  intFormedYear: number;
  strTeamFanart1: string;
  strTeamFanart2: string;
  strTeamFanart3: string;
  strTeamFanart4: string;
}
