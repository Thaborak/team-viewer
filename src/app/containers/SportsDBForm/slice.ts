/*
 * SportsDBForm Slice
 *
 * Here we define:
 * - The shape of our container's slice of Redux store,
 * - All the actions which can be triggered for this slice, including their effects on the store.
 *
 * Note that, while we are using dot notation in our reducer, we are not actually mutating the state
 * manually. Under the hood, we use immer to apply these updates to a new copy of the state.
 *
 */

import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState, TeamErrorType } from './types';
import { Team } from 'types/Team';

// The initial state of the SportsDBForm container
export const initialState: ContainerState = {
  teamname: 'new_york_giants',
  teams: [],
  loading: false,
  error: null,
};

const sportsDbFormSlice = createSlice({
  name: 'sportsDbRepoForm',
  initialState,
  reducers: {
    changeTeamName(state, action: PayloadAction<string>) {
      state.teamname = action.payload;
    },
    loadTeam(state) {
      state.loading = true;
      state.error = null;
      state.teams = [];
    },
    teamLoaded(state, action: PayloadAction<Team[]>) {
      const teams = action.payload;
      state.teams = teams;
      state.loading = false;
    },
    teamError(state, action: PayloadAction<TeamErrorType>) {
      state.error = action.payload;
      state.loading = false;
    },
  },
});

export const { actions, reducer, name: sliceKey } = sportsDbFormSlice;
