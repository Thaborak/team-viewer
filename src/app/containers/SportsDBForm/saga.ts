import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { request } from 'utils/request';
import { selectName } from './selectors';
import { actions } from './slice';
import { Team } from 'types/Team';
import { TeamErrorType } from './types';

/**
 * SportsDB teams request/response handler
 */
export function* getTeam() {
  yield delay(500);
  // Select teamname from store
  const teamname: string = yield select(selectName);
  if (teamname.length === 0) {
    yield put(actions.teamError(TeamErrorType.TEAMNAME_EMPTY));
    return;
  }
  const requestURL = `https://www.thesportsdb.com/api/v1/json/1/searchteams.php?t=${teamname}`;

  try {
    // Call our request helper (see 'utils/request')
    const res = yield call(request, requestURL);
    const teams: Team[] = res.teams;
    if (teams?.length > 0) {
      yield put(actions.teamLoaded(teams));
    } else {
      yield put(actions.teamError(TeamErrorType.TEAM_HAS_NO_DATA));
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.teamError(TeamErrorType.TEAM_NOT_FOUND));
    } else if (err.message === 'Failed to fetch') {
      yield put(actions.teamError(TeamErrorType.SPORTSDB_RATE_LIMIT));
    } else {
      yield put(actions.teamError(TeamErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* sportsDbFormSaga() {
  // Watches for loadRepos actions and calls getTeam when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loadTeam.type, getTeam);
}
