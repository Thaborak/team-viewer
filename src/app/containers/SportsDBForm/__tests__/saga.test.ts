import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '../slice';

import { sportsDbFormSaga, getTeam } from '../saga';
import { TeamErrorType } from '../types';

describe('getTeam Saga', () => {
  let username: any;
  let repos: any;
  let getTeamIterator: ReturnType<typeof getTeam>;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getTeamIterator = getTeam();
    const delayDescriptor = getTeamIterator.next().value;
    expect(delayDescriptor).toMatchSnapshot();

    const selectDescriptor = getTeamIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should return error if username is empty', () => {
    username = '';
    const putDescriptor = getTeamIterator.next(username).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.teamError(TeamErrorType.TEAMNAME_EMPTY)),
    );

    const iteration = getTeamIterator.next();
    expect(iteration.done).toBe(true);
  });

  it('should dispatch the teamsLoaded action if it requests the data successfully', () => {
    const teamname = 'new_york_giants';
    const teams = [
      {
        name: 'New York Giants',
      },
    ];

    const requestDescriptor = getTeamIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getTeamIterator.next(repos).value;
    expect(putDescriptor).toEqual(put(slice.actions.teamLoaded(teams)));
  });

  it('should dispatch the user not found error', () => {
    username = 'test';

    const requestDescriptor = getTeamIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getTeamIterator.throw({ response: { status: 404 } })
      .value;
    expect(putDescriptor).toEqual(
      put(slice.actions.teamError(TeamErrorType.TEAM_NOT_FOUND)),
    );
  });
  it('should dispatch the user has no repo error', () => {
    username = 'test';
    repos = [];

    const requestDescriptor = getTeamIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getTeamIterator.next(repos).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.teamError(TeamErrorType.TEAM_HAS_NO_DATA)),
    );
  });
  it('should dispatch the github rate limit error', () => {
    username = 'test';

    const requestDescriptor = getTeamIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getTeamIterator.throw(new Error('Failed to fetch'))
      .value;
    expect(putDescriptor).toEqual(
      put(slice.actions.teamError(TeamErrorType.SPORTSDB_RATE_LIMIT)),
    );
  });

  it('should dispatch the response error', () => {
    username = 'test';

    const requestDescriptor = getTeamIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getTeamIterator.throw(new Error('some error')).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.teamError(TeamErrorType.RESPONSE_ERROR)),
    );
  });
});

describe('sportsDbFormSaga Saga', () => {
  const apiFormIterator = sportsDbFormSaga();
  it('should start task to watch for loadRepos action', () => {
    const takeLatestDescriptor = apiFormIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.actions.loadTeam.type, getTeam),
    );
  });
});
