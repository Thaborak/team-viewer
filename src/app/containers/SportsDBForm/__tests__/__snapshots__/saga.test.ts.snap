// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTeam Saga should dispatch the github rate limit error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the github rate limit error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTeam Saga should dispatch the github rate limit error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://www.thesportsdb.com/api/v1/json/1/searchteams.php?t=test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTeam Saga should dispatch the response error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://www.thesportsdb.com/api/v1/json/1/searchteams.php?t=test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the teamsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the teamsLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTeam Saga should dispatch the teamsLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 3,
      "type": "sportsDbRepoForm/teamError",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getTeam Saga should dispatch the user has no repo error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the user has no repo error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTeam Saga should dispatch the user has no repo error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://www.thesportsdb.com/api/v1/json/1/searchteams.php?t=test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the user not found error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should dispatch the user not found error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTeam Saga should dispatch the user not found error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://www.thesportsdb.com/api/v1/json/1/searchteams.php?t=test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should return error if username is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTeam Saga should return error if username is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
